CMSgov/easi-app
# [EASI-3179] Retire LCIDs

# EASI-3179

## Changes and Description

- Add GraphQL mutation for retiring an intake's LCID, with a specified retirement date.
- Add GraphQL mutation for changing an intake's LCID retirement date (for use when it already has a retirement date set).
- Store LCID retirement date as a column in the `system_intakes` table and a field on the `SystemIntake` model.
- Update `lcidStatus` field - include logic to return the `RETIRED` status if an LCID's retirement date has passed, as well as moving the code into a method on `SystemIntake`.
- Change the logic for sending LCID expiration alert emails - if an intake has an LCID retirement date set, don't send alert emails for it, even if the retirement date hasn't been reached.
- Add fields to the `actions` table/model/GQL schema to track changes in LCID retirement dates.
- Add Postman requests for both new mutations, under System Intake -> Admin Actions (v2) -> Retire LCID.

## How to test this change
First, import the Postman collection to get the new requests.

For each test:
- Create an intake, submit it, and issue an LCID for it with existing requests.
- Use the "SystemIntake Get (Decision)" query to check an intake's `lcidStatus` and retirement date (`lcidRetiresAt`).
- Use the "SystemIntake Get (Feedback with LCID Change History)" query to get the log of actions for an intake, including changes to the retirement date. 

Major tests:
- Make sure that retiring an LCID and changing an LCID's retirement date update the correct fields to the values from the input - `lcidRetiresAt` on the system intake itself, as well as  `previousRetirementDate`, `newRetirementDate`, and `feedback` on the intake's actions.
- Testing `lcidStatus`:
    - If an intake's retirement date has passed, `lcidStatus` should be `RETIRED`, regardless of whether the expiration date has been passed or is still in the future.
    - If an intake's retirement date is upcoming but its expiration date has already been passed, `lcidStatus` should be `EXPIRED`.
    - If both the retirement date and expiration date are upcoming, `lcidStatus` should be `ISSUED` (meaning that it's active).
- To check the LCID expiration alert logic - use Postman to set up the intakes you want, then run `docker compose restart easi` to restart the backend container and re-run the expiration alert check. Open Mailcatcher at http://localhost:1080/ to check what emails have been sent.

## Notes
- When adding `previousRetirementDate` and `newRetirementDate` to the GraphQL schema, I thought about putting them on the `SystemIntakeLCIDExpirationChange` type, but I didn't like putting non-expiration-related fields on there. The theoretical best solution would probably be to rename `SystemIntakeLCIDExpirationChange` to just something like `SystemIntakeLCIDChange` to reflect that it captures all LCID-related changes, but that renaming would cause a lot of code churn for marginal benefit.
- I removed the comments in `pkg/graph/resolvers/system_intake_actions.go` mentioning `Send Notification Email (EASI-3109)` because @mynar7 added those notification emails in https://github.com/CMSgov/easi-app/pull/2200.
